PLAT-10277: RSA OBO Authentication Implementation

symphony-hong opened the pull request:
### Ticket
[PLAT-10277](https://perzoinc.atlassian.net/browse/PLAT-10277)

### Description
RSA OBO Authentication Implementation.

obo_authenticator.py contains OboAuthenticatorRSA to authenticate the bot using the rsa private key.

OboAuthSession class in auth_session.py to handle a Obo Authentication session.

AuthenticationFactory#get_obo_authenticator method to get new instance of the obo authenticator based on the config.

Fluent method in symphony_bdk.py to perform the obo authentication.

Unittest added.

Docstring added.

### Dependencies
List the other pull requests that should be merged before/along this one.

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Docstrings added or updated
- [ ] Updated the documentation in [docs folder](../docs)

symphony-hong committed:
RSA OBO Authentication Implementation.

obo_authenticator.py contains OboAuthenticatorRSA to authenticate the bot using the rsa private key.

OboAuthSession class in auth_session.py to handle a Obo Authentication session.

AuthenticationFactory#get_obo_authenticator method to get new instance of the obo authenticator based on the config.

Fluent method in symphony_bdk.py to perform the obo authentication.

Unittest added.

Docstring added.

symphony-hong committed:
Resolve conflict upstream/2.0

symphony-hong committed:
Fix import

symphony-hong committed:
Update docstring

symphony-elias commented on symphony/bdk/core/auth/auth_session.py:
```diff
@@ -53,3 +53,42 @@ def key_manager_token(self, value):
+
+    def __init__(self, authenticator, user_id: int = None, username: str = None):
+        """
+        At least user_id or username should be defined.
```
I guess both should not be specified.

symphony-hong replied:
I didn't find the best way to initialise the auth_session here because we cannot have the overloaded method in Python.

symphony-youri replied:
then should we throw an exception if both or none are passed?

symphony-elias commented on symphony/bdk/core/auth/auth_session.py:
```diff
@@ -53,3 +53,42 @@ def key_manager_token(self, value):
+    @property
+    async def session_token(self):
+        if self._session_token is None:
+            if self.user_id is not None:
```
Please reuse `refresh()`:
```python
if self._session_token is None:
  self.refresh()
```

symphony-elias commented on symphony/bdk/core/auth/auth_session.py:
```diff
@@ -53,3 +53,42 @@ def key_manager_token(self, value):
+
+    @property
+    async def key_manager_token(self):
+        return self._key_manager_token
```
This should return `None` or similar.

symphony-hong committed:
Refactor OboAuthSession

symphony-youri commented on symphony/bdk/core/auth/obo_authenticator.py:
```diff
@@ -0,0 +1,145 @@
+        return auth_session
+
+
+class OboAuthenticatorRSA(OboAuthenticator):
```
Rsa to stick to camel case?

symphony-youri commented on symphony/bdk/core/auth/bot_authenticator.py:
```diff
@@ -1,12 +1,13 @@
+
 from symphony.bdk.core.auth.auth_session import AuthSession
 from symphony.bdk.core.auth.exception import AuthUnauthorizedException
+from symphony.bdk.core.auth.jwt_helper import create_signed_jwt
```
in create_signed_jwt we still have this magic 5*58, could we make it a constant and document that JWT needs to have a ttl of 5 minutes max

symphony-youri commented on symphony/bdk/core/symphony_bdk.py:
```diff
@@ -65,6 +65,19 @@ def bot_session(self) -> AuthSession:
         """
         return self._bot_session
 
+    def obo(self, user_id=None, username=None) -> OboAuthSession:
```
should we set the parameter types here?

symphony-hong committed:
Update

Rename RSA -> Rsa to conform the camel case.

Throw exception if the OboAuthSession is constructed with both of username and user id.

Add typing annotation to SymphonyBdk#obo

thibauult commented on tests/resources/config/config.yaml:
```diff
@@ -13,3 +13,8 @@ bot:
   privateKey:
     path: /Users/local/conf/agent/privatekey.pem
+
+app:
+  appId: app-id
+  privateKey:
+    path: /Users/local/conf/agent/privatekey.pem
```
Please also update https://github.com/SymphonyPlatformSolutions/symphony-api-client-python/blob/2.0/docs/configuration.md accordingly :)

symphony-hong committed:
Update documentation

symphony-youri commented on symphony/bdk/core/auth/jwt_helper.py:
```diff
@@ -1,20 +1,23 @@
+    :param private_key_config:  The private key configuration for a service account or an extension app.
+    :param username:            The username of the user to authenticate
+    :param expiration:          Expiration of the authentication request in seconds.
+                                By default the signed jwt will be valid in maximum 290 seconds.
```
thanks, just maybe precise that 5 minutes is the maximum accepted by the backend?
